### PR TITLE
Allow reading claims from CSV for test deployment

### PR DESCRIPTION
**Pull request by fedgiac**
Instead of always generating random claims, allows to specify a CSV claim file that will be used as the list of all claims available in the test deployment.

Note that the plan is to isolate the random claim generation to its own script and use the csv file mode as the only mode of operation of this script. This means that in a future PR all the `if`s introduced in this PR will be removed.

This change requires the [csv npm package](https://www.npmjs.com/package/csv). To my understanding it's a well-known package to manage CSV files.

### Test Plan

New unit tests. Also manual script testing.

First create a file `./claims.csv` with the following content:
```
Account,Airdrop
0x4242424242424242424242424242424242424242,1337
```

Then execute the test deployment script pointing to this file:
```
$ npx hardhat test-deployment --network rinkeby --mnemonic "claim all milk together toward moon swap coin become rich quick guitar" ./claims.csv 
Using deployer 0xe90C3A883c386636192B0199BbDC933a66BDf43c
Reading user claims from file...
Generating Merkle proofs...
Setting up administration addresses...
Generating deploy transactions...
Clearing old files...
Saving generated data to file...
Deploying real token...
Deploying virtual token...
```

Look for example at `output/claims.json` and notice that there is only a single claim, as expected:
```
[{"account":"0x4242424242424242424242424242424242424242","type":0,"claimableAmount":{"type":"BigNumber","hex":"0x0539"},"index":0,"proof":[]}]
```

### CODE REVIEW

**Review by bh2smith — APPROVED** [review]
Looks good. However, it appears you are expecting so see the same account appearing in multiple rows for each different claim type. Is this something you would prefer?

It was my plan to have unique accounts with each claim type as a separate column header.

**Comment by fedgiac** [comment]
> Looks good. However, it appears you are expecting so see the same account appearing in multiple rows for each different claim type. Is this something you would prefer?

No, I personally would like it more if it had only one row for each user with multiple claims (like in [this test](https://github.com/gnosis/cow-token/pull/35/files#diff-ab71404ea11165232742cfd08a3141a12cd737a31ede19894f7f87359590275bR28-R30)). But in principle we might want to have separate claims in different lines, so since it's easier to do I though it was ok to have the same address repeating. Do you think I should rather check that no address repeats and error out in case?

> It was my plan to have unique accounts with each claim type as a separate column header.

:+1:

**Comment by bh2smith** [comment]
I can do either way (or both). At the moment the plan for user digestibility is to have one row for each account, but I will just generate two files at the end.

File: src/tasks/test-deployment.ts [hunk 1]
```diff
@@ -83,11 +86,11 @@ interface CleanArgs {
...
-  gnosisDaoAddress?: string;
-  cowDaoAddress?: string;
-  communityFundsTargetAddress?: string;
-  investorFundsTargetAddress?: string;
-  teamControllerAddress?: string;
+  gnosisDaoAddress: string | undefined;
+  cowDaoAddress: string | undefined;
+  communityFundsTargetAddress: string | undefined;
+  investorFundsTargetAddress: string | undefined;
+  teamControllerAddress: string | undefined;
```
Comment (anxolin) [inline src/tasks/test-deployment.ts:93]: why these changes?

**Review by anxolin — COMMENTED** [review]
It worked pretty well with the exception of one address. I didn't see it in the chunks outputs (`0x60535AFCB553046DfB3636dc04A9cd159Bd652F1`):

CSV I used:
```csv
 Account,Airdrop,GnoOption,UserOption,Investor,TeamAdvisor
  2 0x424a46612794dbb8000194937834250Dc723fFa5,1000000000000000000000,0,0,0,0
  3 0x1AF7b61b4Fbb44495d3c55f402c06567c619E21C,0,1500000000000000000000,0,0,0
  4 0x60535AFCB553046DfB3636dc04A9cd159Bd652F1,0,0,0,0,1500000000000000000000
  5 0x0652572f107bc41D294cDe6Be09d50791bafed02,1500000000000000000000,1500000000000000000000,1500000000000000000000,1500000000000000000000,1500000000000000000000
```

Output: 
```
Using deployer 0x4c43edcb55877b02113B2afB66D7325a2e39DDD6
Reading user claims from file...
Generating Merkle proofs...
Setting up administration addresses...
Generating deploy transactions...
Clearing old files...
Saving generated data to file...
Deploying real token...
Deploying virtual token...
```


The only chunk generated:
```json
[
  {
    "account": "0x424a46612794dbb8000194937834250Dc723fFa5",
    "type": 0,
    "claimableAmount": {
      "type": "BigNumber",
      "hex": "0x3635c9adc5dea00000"
    },
    "index": 5,
    "proof": [
      "0x2e83dc03124ce66513bdabc063004ac9e454cd6f8cb141e7e3fab5d65af9bd6b",
      "0x4204482a090573e1fca272aa4c22c6ae583dc4beff30388f6aedffffbfa9dcaf",
      "0x82fcffb3eb91c11674d822209b8d39649aa5ad498f00d18c9106938dde1d27aa"
    ]
  },
  {
    "account": "0x1AF7b61b4Fbb44495d3c55f402c06567c619E21C",
    "type": 1,
    "claimableAmount": {
      "type": "BigNumber",
      "hex": "0x5150ae84a8cdf00000"
    },
    "index": 4,
    "proof": [
      "0x049315aedae6553374e7d41cf35eac9a2f969f1807cd960a2d517be61b4d35cf",
      "0x3785f98937f2f907ea14f10114e249e33a4b06b30f9ba59003ecc73c211a8971",
      "0x82fcffb3eb91c11674d822209b8d39649aa5ad498f00d18c9106938dde1d27aa"
    ]
  },
  {
    "account": "0x0652572f107bc41D294cDe6Be09d50791bafed02",
    "type": 0,
    "claimableAmount": {
      "type": "BigNumber",
      "hex": "0x5150ae84a8cdf00000"
    },
    "index": 0,
    "proof": [
      "0x469785d363e21495d7f22acbb9b54f5cdde08cd6ad8e7c723625d66d6e89b310",
      "0x46c5efc7f17e7cf344ea338a8ef345c8d9b61adf6bf3e012aa2fc1ed0756d144"
    ]
  },
  {
    "account": "0x0652572f107bc41D294cDe6Be09d50791bafed02",
    "type": 1,
    "claimableAmount": {
      "type": "BigNumber",
      "hex": "0x5150ae84a8cdf00000"
    },
    "index": 1,
    "proof": [
      "0x00ac7d9eda80339218ce8d1a10762ca33b072c956cedf6c41fd8fdb80d555b53",
      "0x3785f98937f2f907ea14f10114e249e33a4b06b30f9ba59003ecc73c211a8971",
      "0x82fcffb3eb91c11674d822209b8d39649aa5ad498f00d18c9106938dde1d27aa"
    ]
  },
  {
    "account": "0x0652572f107bc41D294cDe6Be09d50791bafed02",
    "type": 2,
    "claimableAmount": {
      "type": "BigNumber",
      "hex": "0x5150ae84a8cdf00000"
    },
    "index": 2,
    "proof": [
      "0x31b6369a83ee73990effa2adecb1e74a5901431694bcb79f9dd31a58381e63fe",
      "0x4204482a090573e1fca272aa4c22c6ae583dc4beff30388f6aedffffbfa9dcaf",
      "0x82fcffb3eb91c11674d822209b8d39649aa5ad498f00d18c9106938dde1d27aa"
    ]
  },
  {
    "account": "0x0652572f107bc41D294cDe6Be09d50791bafed02",
    "type": 3,
    "claimableAmount": {
      "type": "BigNumber",
      "hex": "0x5150ae84a8cdf00000"
    },
    "index": 3,
    "proof": [
      "0x4a89f70a1df62ada8593c19bd66eedd1e648e477171bfa333633e2a306146b99",
      "0x46c5efc7f17e7cf344ea338a8ef345c8d9b61adf6bf3e012aa2fc1ed0756d144"
    ]
  }
]
```

**Comment by fedgiac** [comment]
I believe the issue is in the CSV. The header is:
```
 Account,Airdrop,GnoOption,UserOption,Investor,TeamAdvisor
```
when it should be:
```
 Account,Airdrop,GnoOption,UserOption,Investor,Team,Advisor
```

So what happened is that the script found a column with an unknown header and it was ignored. `0x60535AFCB553046DfB3636dc04A9cd159Bd652F1` has a claim only for the last column, so it doesn't appear.

I think skipping unknown fields is reasonable. Do you think the script should enforce all expected headers to be there?

**Review by josojo — APPROVED** [review]
Works smoothly!

(I thought I would have the same issue as anxo and that by just adding a line break at the end of the file fixed it. But when Federico asked me to replicate it, I couldn't)

File: src/tasks/test-deployment.ts (hunk 1 above)
Comment (fedgiac) [reply src/tasks/test-deployment.ts:93]: I was adding `claimCsv` to the same interface, and that field has the same restrictions as these. I realized that these fields were never really optional and were always there, but just be set to `undefined` sometimes. So I fixed all the types here since it's a very minor change.

File: src/tasks/test-deployment.ts [hunk 2]
```diff
@@ -58,6 +59,7 @@ const defaultArgs = {
 } as const;
 interface DeployTaskArgs {
   mnemonic: string;
+  claimCsv?: string;
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply src/tasks/test-deployment.ts:62]: This could actually be not existing as it comes from Hardhat. Note that it's a different type (`DeployTaskArgs`) to the one with the changes from `?:` to `: string | undefined` (`CleanArgs`).

File: src/tasks/test-deployment.ts (hunk 1 above)
Comment (anxolin) [reply src/tasks/test-deployment.ts:93]: okok, i thought it was equivalent and preferable the `?` form, but i guess u want to signal its presence even if they are undefined. NVM, just curious

**Comment by anxolin** [comment]
> I'm not able to replicate this, it work for me without line breaks. Can you send me the CSV?

Did it work with my example of CSV?

**Comment by fedgiac** [comment]
Note that your CSV has some spurious numbers to the left of the addresses which I removed for the test, this is what I used to see the issue you mentioned:
```
Account,Airdrop,GnoOption,UserOption,Investor,TeamAdvisor
0x424a46612794dbb8000194937834250Dc723fFa5,1000000000000000000000,0,0,0,0
0x1AF7b61b4Fbb44495d3c55f402c06567c619E21C,0,1500000000000000000000,0,0,0
0x60535AFCB553046DfB3636dc04A9cd159Bd652F1,0,0,0,0,1500000000000000000000
0x0652572f107bc41D294cDe6Be09d50791bafed02,1500000000000000000000,1500000000000000000000,1500000000000000000000,1500000000000000000000,1500000000000000000000
```

Then the issue with this file was discussed [here](https://github.com/gnosis/cow-token/pull/35#issuecomment-1016829364), I changed that file accordingly and I got the script working as expected with the expected number of claims. This is the fixed CSV:
```

Account,Airdrop,GnoOption,UserOption,Investor,Team,Advisor
0x424a46612794dbb8000194937834250Dc723fFa5,1000000000000000000000,0,0,0,0,0
0x1AF7b61b4Fbb44495d3c55f402c06567c619E21C,0,1500000000000000000000,0,0,0,0
0x60535AFCB553046DfB3636dc04A9cd159Bd652F1,0,0,0,0,1500000000000000000000,0
0x0652572f107bc41D294cDe6Be09d50791bafed02,1500000000000000000000,1500000000000000000000,1500000000000000000000,1500000000000000000000,1500000000000000000000,1500000000000000000000
```

**Review by anxolin — APPROVED** [review]
@fedgiac as we just talked, its all good. As you helped me to understand, my example had one unknown column, `TeamAdvisor` and the script ignore unknown columns

All good from my side. We just need to be careful with the column headers.
One nitpick would be to print a warning with the ignored columns.

**Comment by fedgiac** [comment]
>  One nitpick would be to print a warning with the ignored columns.

Makes sense! I'll add a warning in a follow-up PR.
